# Loan Prediction

Loan prediction for customers using  logistic Regression,decision tree model and random forest model.


In [49]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.metrics import f1_score
from sklearn.model_selection import train_test_split

 Importing dataset

In [50]:
df=pd.read_csv('loan_train.csv')
df.head()

Loan_ID Gender Married Dependents     Education Self_Employed  \
0  LP001002   Male      No          0      Graduate            No   
1  LP001003   Male     Yes          1      Graduate            No   
2  LP001005   Male     Yes          0      Graduate           Yes   
3  LP001006   Male     Yes          0  Not Graduate            No   
4  LP001008   Male      No          0      Graduate            No   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849                0.0         NaN             360.0   
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   

   Credit_History Property_Area Loan_Status  
0             1.0         Urban           Y  
1             1.0         Rural           N  
2             1.0         Urban           Y  
3             1.0         Urban           Y  
4             1.0         Urban           Y

In [51]:
df['Gender'] = df['Gender'].fillna( df['Gender'].dropna().mode().values[0] )
df['Married'] = df['Married'].fillna( df['Married'].dropna().mode().values[0] )
df['Dependents'] = df['Dependents'].fillna( df['Dependents'].dropna().mode().values[0] )
df['Self_Employed'] = df['Self_Employed'].fillna( df['Self_Employed'].dropna().mode().values[0] )
df['LoanAmount'] = df['LoanAmount'].fillna( df['LoanAmount'].dropna().mean() )
df['Loan_Amount_Term'] = df['Loan_Amount_Term'].fillna( df['Loan_Amount_Term'].dropna().mode().values[0] )
df['Credit_History'] = df['Credit_History'].fillna( df['Credit_History'].dropna().mode().values[0] )
df['Dependents'].replace('3+',4,inplace=True)

In [52]:
df['Gender']=df['Gender'].map({'Male':1,'Female':0})
df['Married']=df['Married'].map({'Yes':1,'No':0})
df['Education']=df['Education'].map({'Graduate':1,'Not Graduate':0})
df['Dependents'].replace('3+',4,inplace=True)
df['Self_Employed']=df['Self_Employed'].map({'Yes':1,'No':0})
df['Property_Area']=df['Property_Area'].map({'Semiurban':1,'Urban':2,'Rural':3})
df['Loan_Status']=df['Loan_Status'].map({'Y':1,'N':0})


In [53]:
df.dtypes

Loan_ID               object
Gender                 int64
Married                int64
Dependents            object
Education              int64
Self_Employed          int64
ApplicantIncome        int64
CoapplicantIncome    float64
LoanAmount           float64
Loan_Amount_Term     float64
Credit_History       float64
Property_Area          int64
Loan_Status            int64
dtype: object

In [54]:
df.head()

Loan_ID  Gender  Married Dependents  Education  Self_Employed  \
0  LP001002       1        0          0          1              0   
1  LP001003       1        1          1          1              0   
2  LP001005       1        1          0          1              1   
3  LP001006       1        1          0          0              0   
4  LP001008       1        0          0          1              0   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849                0.0  146.412162             360.0   
1             4583             1508.0  128.000000             360.0   
2             3000                0.0   66.000000             360.0   
3             2583             2358.0  120.000000             360.0   
4             6000                0.0  141.000000             360.0   

   Credit_History  Property_Area  Loan_Status  
0             1.0              2            1  
1             1.0              3            0  
2             1.0              2            1  
3             1.0              2            1  
4             1.0              2            1

Machine Learning Algorithm

1.Logistic Regression.

2.Decision Tree Classifier.

3.Random Forest Classifier.

1.Logistic Regression

In [55]:
y=df.Loan_Status
df_V1=df.drop('Loan_Status',1)

C:\Users\TOSHIBA\AppData\Local\Temp/ipykernel_2228/1545960223.py:2: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  df_V1=df.drop('Loan_Status',1)


In [56]:
df_V1=pd.get_dummies(df_V1)
df=pd.get_dummies(df)

In [57]:
df_V2 = df.filter(['Credit_History'], axis=1)
df_V2=pd.get_dummies(df_V2)

In [58]:
import statsmodels.api as sm
from statsmodels.sandbox.regression.predstd import wls_prediction_std
from sklearn.metrics import accuracy_score


In [59]:
model2=sm.Logit(y,df_V2 )

result2=model2.fit()

print(result2.summary())

Optimization terminated successfully.
         Current function value: 0.539390
         Iterations 5
                           Logit Regression Results                           
Dep. Variable:            Loan_Status   No. Observations:                  614
Model:                          Logit   Df Residuals:                      613
Method:                           MLE   Df Model:                            0
Date:                Thu, 09 Jun 2022   Pseudo R-squ.:                  0.1318
Time:                        13:30:21   Log-Likelihood:                -331.19
converged:                       True   LL-Null:                       -381.45
Covariance Type:            nonrobust   LLR p-value:                       nan
                     coef    std err          z      P>|z|      [0.025      0.975]
----------------------------------------------------------------------------------
Credit_History     1.3278      0.107     12.381      0.000       1.118       1.538


In [60]:
from sklearn.linear_model import LogisticRegression

model_V2=LogisticRegression()
model_V2.fit(df_V2,y)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True, intercept_scaling=1, max_iter=100,multi_class='ovr',n_jobs=1,penalty='12',random_state=1,solver='liblinear',tol=0.0001,verbose=0,warm_start=False)

d=pd.DataFrame({'odds_ratio':(np.exp(model_V2.coef_).T).tolist(),'variable':df_V2.columns.tolist()})
d['odds_ratio'] = d['odds_ratio'].str.get(0)

d=d.sort_values('odds_ratio', ascending=False)
d

odds_ratio        variable
0    27.69363  Credit_History

In [61]:
from sklearn import metrics
predicted_2 = model_V2.predict(df_V2)
metrics.accuracy_score(y,predicted_2)

0.8094462540716613

Creating third model with two columns credit history and coapplicant income.

In [62]:
df_V3 = df1.filter(['Credit_History', 'CoapplicantIncome'], axis=1)
df_V3=pd.get_dummies(df_V3)

In [63]:
model3=sm.Logit(y,df_V3)

result3=model3.fit()

print(result3.summary())

Optimization terminated successfully.
         Current function value: 0.530149
         Iterations 5
                           Logit Regression Results                           
Dep. Variable:            Loan_Status   No. Observations:                  614
Model:                          Logit   Df Residuals:                      612
Method:                           MLE   Df Model:                            1
Date:                Thu, 09 Jun 2022   Pseudo R-squ.:                  0.1466
Time:                        13:30:43   Log-Likelihood:                -325.51
converged:                       True   LL-Null:                       -381.45
Covariance Type:            nonrobust   LLR p-value:                 3.818e-26
                        coef    std err          z      P>|z|      [0.025      0.975]
-------------------------------------------------------------------------------------
Credit_History        1.5175      0.126     12.004      0.000       1.270       1.765
Coapplic

In [64]:
model_V3 = LogisticRegression()
model_V3.fit(df_V3,y)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True, intercept_scaling=1, max_iter=100,multi_class='ovr',n_jobs=1,penalty='12',random_state=1,solver='liblinear',tol=0.0001,verbose=0,warm_start=False)

df=pd.DataFrame({'odds_ratio':(np.exp(model_V3.coef_).T).tolist(),'variable':df_V3.columns.tolist()})
df['odds_ratio'] = df['odds_ratio'].str.get(0)

df=df.sort_values('odds_ratio', ascending=False)
df

odds_ratio           variable
0   28.048905     Credit_History
1    0.999948  CoapplicantIncome

In [65]:
predicted_3 = model_V3.predict(df_V3)
accuracy_score(y,predicted_3)

0.8127035830618893

In [ ]:
Testing the model by splitting data into 80% for train and 20 % for test.

In [67]:
from sklearn.linear_model import LogisticRegression
from sklearn import metrics
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(df_V3, y, test_size=0.2, random_state=0)
logreg = LogisticRegression()
logreg.fit(X_train, y_train)

LogisticRegression()

In [68]:
y_pred = logreg.predict(X_test)
print('Accuracy of logistic regression classifier on test set: {:.2f}'.format(logreg.score(X_test, y_test)))


Accuracy of logistic regression classifier on test set: 0.84


Dummy variables  created  for constructing the model.

Finally,the model with coapplicant income and credit history as independent variable with highest accuracy.

The Tested data tested  got the accuracy of 84%.


In [69]:
df=pd.read_csv('loan_train.csv')
df.head()

Loan_ID Gender Married Dependents     Education Self_Employed  \
0  LP001002   Male      No          0      Graduate            No   
1  LP001003   Male     Yes          1      Graduate            No   
2  LP001005   Male     Yes          0      Graduate           Yes   
3  LP001006   Male     Yes          0  Not Graduate            No   
4  LP001008   Male      No          0      Graduate            No   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849                0.0         NaN             360.0   
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   

   Credit_History Property_Area Loan_Status  
0             1.0         Urban           Y  
1             1.0         Rural           N  
2             1.0         Urban           Y  
3             1.0         Urban           Y  
4             1.0         Urban           Y

In [70]:
df['Gender'] = df['Gender'].fillna( df['Gender'].dropna().mode().values[0] )
df['Married'] = df['Married'].fillna( df['Married'].dropna().mode().values[0] )
df['Dependents'] = df['Dependents'].fillna( df['Dependents'].dropna().mode().values[0] )
df['Self_Employed'] = df['Self_Employed'].fillna( df['Self_Employed'].dropna().mode().values[0] )
df['LoanAmount'] = df['LoanAmount'].fillna( df['LoanAmount'].dropna().mean() )
df['Loan_Amount_Term'] = df['Loan_Amount_Term'].fillna( df['Loan_Amount_Term'].dropna().mode().values[0] )
df['Credit_History'] = df['Credit_History'].fillna( df['Credit_History'].dropna().mode().values[0] )
df['Dependents'].replace('3+',4,inplace=True)

In [71]:
df['Gender']=df['Gender'].map({'Male':1,'Female':0})
df['Married']=df['Married'].map({'Yes':1,'No':0})
df['Education']=df['Education'].map({'Graduate':1,'Not Graduate':0})
df['Dependents'].replace('3+',4,inplace=True)
df['Self_Employed']=df['Self_Employed'].map({'Yes':1,'No':0})
df['Property_Area']=df['Property_Area'].map({'Semiurban':1,'Urban':2,'Rural':3})
df['Loan_Status']=df['Loan_Status'].map({'Y':1,'N':0})

In [72]:
df.dtypes

Loan_ID               object
Gender                 int64
Married                int64
Dependents            object
Education              int64
Self_Employed          int64
ApplicantIncome        int64
CoapplicantIncome    float64
LoanAmount           float64
Loan_Amount_Term     float64
Credit_History       float64
Property_Area          int64
Loan_Status            int64
dtype: object

2.Decision Tree Classifier

Testing the model by splitting data into 80% for train and 20 % for test.

In [73]:
X=df.drop(columns=['Loan_ID','Loan_Status']).values
Y=df['Loan_Status'].values
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, random_state = 0)


In [74]:
print('Shape of X_train=>',X_train.shape)
print('Shape of X_test=>',X_test.shape)
print('Shape of Y_train=>',Y_train.shape)
print('Shape of Y_test=>',Y_test.shape)

Shape of X_train=> (491, 11)
Shape of X_test=> (123, 11)
Shape of Y_train=> (491,)
Shape of Y_test=> (123,)


Building and Evaluating the Model with Decision Tree

Training the models and classifing the loan applications.

Training the decision tree on this datase.

Next, To evaluate this model using F1-Score.

In [75]:
from sklearn.tree import DecisionTreeClassifier
dt = DecisionTreeClassifier(criterion = 'entropy', random_state = 0)
dt.fit(X_train, Y_train)
dt_pred_train = dt.predict(X_train)

Evaluatiion on Training set

In [76]:
dt_pred_train = dt.predict(X_train)
print('Training Set Evaluation with  Decision Tree F1-Score=>',f1_score(Y_train,dt_pred_train))


Training Set Evaluation with  Decision Tree F1-Score=> 1.0


Evaluation on Testing set

In [77]:
dt_pred_test = dt.predict(X_test)
print('Testing Set Evaluation with  Decision Tree F1-Score=>',f1_score(Y_test,dt_pred_test))


Testing Set Evaluation with  Decision Tree F1-Score=> 0.8114285714285714


Decision Tree performs well on in-sample evaluation, but its performance decreases drastically on out-of-sample evaluation.


3.RandomForestClassifier

In [78]:
from sklearn.ensemble import RandomForestClassifier
rfc = RandomForestClassifier(criterion = 'entropy', random_state = 0)
rfc.fit(X_train, Y_train)

RandomForestClassifier(criterion='entropy', random_state=0)

Evaluating on Training set

In [79]:
rfc_pred_train = rfc.predict(X_train)
print('Training Set Evaluation with  Random Forest F1-Score=>',f1_score(Y_train,rfc_pred_train))


Training Set Evaluation with  Random Forest F1-Score=> 1.0


Evaluating on Test set

In [80]:
rfc_pred_test = rfc.predict(X_test)
print('Testing Set Evaluation with  Random Forest F1-Score=>',f1_score(Y_test,rfc_pred_test))


Testing Set Evaluation with  Random Forest F1-Score=> 0.8571428571428572


Conclusion:

The  Decision Tree model the accuracy score 81%.

The logistic regression model the accuracy score 84%.

The Random Forestmodel accuracy score 85%.

The random forest model performed much better than the decision tree and logistic regression.